pkg/secrets/pushtofile/secrets_groups.go

```diff
@@ -0,0 +1,220 @@
+type SecretsPaths map[string]string
+
+// SecretsGroupInfo comprises secrets mapping information for a given secrets
+// group.
+type SecretsGroupInfo struct {
+	Secrets      SecretsPaths
+	FilePath     string
+	FileFormat   SecretsFileFormat
+	FilePerms    os.FileMode
+	FileTemplate string
+}
+
+// SecretsGroups comprises secrets mapping info for all secret groups
+type SecretsGroups map[string]SecretsGroupInfo
+
+func ExtractSecretsGroupsFromAnnotations(annotations map[string]string) (SecretsGroups, error) {
+	secretsGroups := make(SecretsGroups)
+
+	for annotation := range annotations {
+		if strings.HasPrefix(annotation, CONJUR_SECRETS_PREFIX) {
+			groupName := strings.TrimPrefix(annotation, CONJUR_SECRETS_PREFIX)
+
+			secretsPathPrefix, err := parseConjurSecretsPathPrefix(groupName, annotations[CONJUR_SECRETS_POLICY_PATH_PREFIX+groupName])
+			if err != nil {
+				return nil, err
+			}
+
+			secrets, err := parseConjurSecretsPaths(groupName, annotations[CONJUR_SECRETS_PREFIX+groupName], secretsPathPrefix)
+			if err != nil {
+				return nil, err
+			}
+
+			filePath, err := parseFilePath(groupName, annotations[SECRET_FILE_PATH_PREFIX+groupName])
+			if err != nil {
+				return nil, err
+			}
+
+			fileFormat, err := parseFileFormat(groupName, annotations[SECRET_FILE_FORMAT_PREFIX+groupName])
+			if err != nil {
+				return nil, err
+			}
+
+			fileTemplate := annotations[SECRET_FILE_TEMPLATE_PREFIX+groupName]
+
+			if fileTemplate != "" {
+				fileFormat = FILE_FORMAT_NONE
+			}
+
+			groupInfo := SecretsGroupInfo{
+				Secrets:      secrets,
+				FilePath:     filePath,
+				FileFormat:   fileFormat,
+				FilePerms:    DEFAULT_FILE_PERMS,
+				FileTemplate: fileTemplate,
+			}
+
+			err = validateGroupInfo(groupName, groupInfo)
+			if err != nil {
+				return nil, err
+			}
+
+			secretsGroups[groupName] = groupInfo
+		}
+	}
+
+	return secretsGroups, nil
+}
+
+func parseConjurSecretsPaths(groupName string, secretsPaths string, secretsPathsPrefix string) (SecretsPaths, error) {
+	secrets := make(SecretsPaths)
+
+	unpacked := []interface{}{}
+	if err := yaml.Unmarshal([]byte(secretsPaths), &unpacked); err != nil {
+		return nil, log.RecordedError(messages.CSPFK050E, err.Error(), groupName)
+	}
+
+	insertSecret := func(name string, secretPath string) error {
+		secretPath = path.Clean(secretPath)
+		if secretPath == "." {
+			return log.RecordedError(messages.CSPFK050E, "Invalid secret path", groupName)
+		}
+		secrets[name] = secretsPathsPrefix + secretPath
+
+		return nil
+	}
+
+	for _, secret := range unpacked {
+		switch val := secret.(type) {
+		case string:
+			{
+				name := path.Base(val)
+				if name == "" {
+					return nil, log.RecordedError(messages.CSPFK050E, "Invalid secret name", groupName)
+				}
+
+				if err := insertSecret(name, val); err != nil {
+					return nil, err
+				}
+			}
+
+		case map[string]interface{}:
+			{
+				for alias, secretPath := range val {
+					secretPath, _ := secretPath.(string)
```

Comment:
The case `map[string]interface{}` would match for values of any type (not only strings). Right now we just ignore if the type casting fails. This can happen If the user somehow provides values that are not strings. We would ideally want to inform them through a validation error.

Reply:
The error case was handled but I dropped all this code in the interest of time. IMO: custom unmarshaller is unnecessary here - I prefer standard unmarshalling to `interface{}` type + type switch. Custom unmarshaller requires 2 calls to `unmarshal()` in the worst case. This approach seems more straightforward. Again, changed since I don't want to spend more time on a 0 value add discussion

Reply:
I agree that the implementation of unmarshalling that avoids the 2 calls to `unmarshal` is more straightforward. My suggestion was to move it out into the custom unmarshaller method for the dual benefit of unmarshalling via a "native" API and collocating the unmarshalling logic with the entity that needs to be unmarshalled.

Reply:
I think a custom unmarshaller is entirely unnecessary in this case. There's no need to create an entity just to unmarshal to, especially a struct with 2 fields representing key and value. It should be unmarshalled into a temporary object and then directly put into the appropriate structure, which in this case is a map.